packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -454,25 +472,56 @@ export function buildImportAnalysisPlugin(config: ResolvedConfig): Plugin {
               }
 
               if (markerStartPos > 0) {
-                s.overwrite(
-                  markerStartPos,
-                  markerStartPos + preloadMarkerWithQuote.length,
-                  // the dep list includes the main chunk, so only need to reload when there are
-                  // actual other deps. Don't include the assets dir if the default asset file names
-                  // are used, the path will be reconstructed by the import preload helper
+                const { modulePreload } = config.build
+                // the dep list includes the main chunk, so only need to reload when there are actual other deps.
+                const depsArray =
                   deps.size > 1 ||
-                    // main chunk is removed
-                    (hasRemovedPureCssChunk && deps.size > 0)
-                    ? `[${[...deps]
-                        .map((d) =>
-                          JSON.stringify(
-                            relativePreloadUrls
-                              ? path.relative(path.dirname(file), d)
-                              : d
+                  // main chunk is removed
+                  (hasRemovedPureCssChunk && deps.size > 0)
+                    ? [...deps]
+                    : []
+                const toRelative = (dep: string) =>
+                  path.relative(path.dirname(file), dep)
+                const resolvedDependencies =
+                  url &&
+                  typeof modulePreload === 'object' &&
+                  modulePreload.resolveDependencies
+                    ? modulePreload
+                        .resolveDependencies(url, depsArray, { importer: file })
+                        .map((dep) => {
+                          if (typeof dep === 'object') {
+                            if (dep.runtime) {
+                              return dep.runtime
+                            }
+                            if (dep.relative) {
+                              return JSON.stringify(toRelative(dep.relative))
+                            }
+                            throw new Error(
+                              `Invalid dependency object, no runtime or relative option specified`
+                            )
+                          }
```

Comment:
This seems to be a very advanced feature, especially the `runtime` part. I'm curious how Nuxt will be using this part of the API. Do they have their own runtime preload list?

IIUC this PR would also disable the optimization in #9491, or perhaps doesn't work only if `resolveDependencies` is used. But I don't think it's a big issue since most won't be using this feature directly.

The rest looks great though.

Reply:
Nuxt is using `runtime` to use its own function to resolve paths, like `globalThis.__resolvePath(...)`.

I commented in #9491, that we should wait a bit to avoid breaking the ecosystem but I think we can still do the optimization after this PR.

The PR changed quite a bit after [`9f1eaa5` (#9938)](https://github.com/vitejs/vite/pull/9938/commits/9f1eaa5f7539eb7588506419465aa8c7251b2efd)